### PR TITLE
Doc/improve label connectivity and voxelize binary mask

### PR DESCRIPTION
**Pull request by GuillaumeBroggi**
Improve documentation.

See:
- https://github.com/pyvista/pyvista/pull/6659#issuecomment-2374179001
- https://github.com/pyvista/pyvista/issues/6677

### CODE REVIEW

File: pyvista/core/filters/poly_data.py [hunk 1]
```diff
@@ -3938,6 +3939,8 @@ def voxelize_binary_mask(
         used by default to estimate the spacing. On systems with VTK < 9.2, the default
         spacing is set to ``mesh_length_fraction=1/100``.
 
+        .. versionadded:: 0.45.0
```
Comment (GuillaumeBroggi) [inline pyvista/core/filters/poly_data.py:3942]: To be checked.

File: pyvista/core/filters/poly_data.py [hunk 2]
```diff
@@ -3915,8 +3915,9 @@ def voxelize_binary_mask(
     ):
         """Voxelize :class:`~pyvista.PolyData` as a binary :class:`~pyvista.ImageData` mask.
 
-        Generate a voxelized representation of a surface. The voxelization can be
-        controlled in several ways.
+        Generate a voxelized representation of a surface where the voxels enclosed by the
+        surface are referred to as foreground. The voxelization can be controlled in several
+        ways.
```
Comment (GuillaumeBroggi) [inline pyvista/core/filters/poly_data.py:3920]: Should we describe how a point and/or a voxel is considered background or foreground?

Comment (user27182) [reply pyvista/core/filters/poly_data.py:3920]: > Should we describe how a point and/or a voxel is considered background or foreground?

Probably a good idea given the new examples, which are interesting cases that raise questions about how the filter works.

```
Generate a voxelized representation of a surface where the voxels enclosed by the
surface are referred to as foreground. The voxelization can be controlled in several
ways.
```

Might be good to include more details about what is meant by "enclosed". The specifics on how the algorithm works are detailed in a [source code comment](https://github.com/Kitware/VTK/blob/5fd6ec84fee1af9c4171fdee2a107ea0e49f5f1b/Imaging/Stencil/vtkPolyDataToImageStencil.cxx#L441-L448):

>> // Description of algorithm:
>>  // 1) cut the polydata at each z slice to create polylines
>>  // 2) find all "loose ends" and connect them to make polygons
>>  //    (if the input polydata is closed, there will be no loose ends)
>>  // 3) go through all line segments, and for each integer y value on
>>  //    a line segment, store the x value at that point in a bucket
>>  // 4) for each z integer index, find all the stored x values
>>  //    and use them to create one z slice of the vtkStencilData

There are also helpful comments in the implementation. Not sure what level of detail to include...

File: pyvista/core/filters/poly_data.py [hunk 3]
```diff
@@ -4134,6 +4143,64 @@ def voxelize_binary_mask(
...
+        >>> binary_mask_1["mask"] = (
+        ...     binary_mask_1["mask"] | binary_mask_2["mask"]
+        ... )
+
+        >>> plot = pv.Plotter()
+        >>> _ = plot.add_mesh(binary_mask_1)
+        >>> _ = plot.add_mesh(cylinder_1.slice(), color='red')
+        >>> _ = plot.add_mesh(cylinder_2.slice(), color='red')
+        >>> plot.show(jupyter_backend="notebook", cpos="yz")
+
```
Comment (user27182) [inline pyvista/core/filters/poly_data.py:4203]: I think it's a bit confusing to go from cylinders to disks then back to cylinders. Suggest (1) swapping the two examples (2) adding a note "Note how the intersection is excluded from the mask" and (3) remove `jupyter_backend="notebook"` kwarg from `show()`
```suggestion
        Note how the intersection is excluded from the mask.
        To include the voxels delimited by internal surfaces in the foreground, the internal
        surfaces should be removed, for instance by applying a boolean union. Note that
        this operation in unreliable in VTK but may be performed with external tools such
        as `vtkbool <https://github.com/zippy84/vtkbool>`_.

        Alternatively, the intersecting parts of the mesh can be processed sequentially.

        >>> cylinder_1 = pv.Cylinder()
        >>> cylinder_2 = pv.Cylinder((0, 0.75, 0))

        >>> reference_volume = pv.ImageData(
        ...     dimensions=(1, 100, 50),
        ...     spacing=(1, 0.0175, 0.02),
        ...     origin=(0, -0.5 + 0.0175 / 2, -0.5 + 0.02 / 2),
        ... )

        >>> binary_mask_1 = cylinder_1.voxelize_binary_mask(
        ...     reference_volume=reference_volume
        ... ).points_to_cells()
        >>> binary_mask_2 = cylinder_2.voxelize_binary_mask(
        ...     reference_volume=reference_volume
        ... ).points_to_cells()

        >>> binary_mask_1["mask"] = (
        ...     binary_mask_1["mask"] | binary_mask_2["mask"]
        ... )

        >>> plot = pv.Plotter()
        >>> _ = plot.add_mesh(binary_mask_1)
        >>> _ = plot.add_mesh(cylinder_1.slice(), color='red')
        >>> _ = plot.add_mesh(cylinder_2.slice(), color='red')
        >>> plot.show(cpos="yz")

        When multiple internal surfaces are nested, they are successively treated as
        interfaces between background and foreground.

        >>> mesh = (
        ...     pv.Tube(radius=2) + pv.Tube(radius=3) + pv.Tube(radius=4)
        ... )
        >>> binary_mask = mesh.voxelize_binary_mask(
        ...     dimensions=(1, 50, 50)
        ... ).points_to_cells()
        >>> plot = pv.Plotter()
        >>> _ = plot.add_mesh(binary_mask)
        >>> _ = plot.add_mesh(mesh.slice(), color='red')
        >>> plot.show(cpos="yz")

```

**Comment by user27182** [comment]
LGTM with or without any of the suggested changes. Just needs to resolve a couple of typing errors to pass CI

**Comment by GuillaumeBroggi** [comment]
Thanks for the comments. I am quite busy these days, but should be able to review them by the end of the week.

File: pyvista/core/filters/poly_data.py (hunk 2 above)
Comment (GuillaumeBroggi) [reply pyvista/core/filters/poly_data.py]: See the proposed description.

**Review by user27182 — APPROVED** [review]
LGTM, thanks!

**Pull request merged**